flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputProcessor.java

```diff
@@ -158,43 +173,34 @@ public boolean isFinished() {
 	}
 
 	@Override
-	public boolean processInput() throws Exception {
-		if (!isPrepared) {
-			// the preparations here are not placed in the constructor because all work in it
-			// must be executed after all operators are opened.
-			prepareForProcessing();
-		}
+	public InputStatus processInput() throws Exception {
+		// the preparations here are not placed in the constructor because all work in it
+		// must be executed after all operators are opened.
+		prepareForProcessing();
 
 		int readingInputIndex = selectNextReadingInputIndex();
 		if (readingInputIndex == -1) {
-			return false;
+			return InputStatus.NOTHING_AVAILABLE;
```

Comment:
Should this clause actually check `firstInputStatus` and `secondInputStatus` that they are both `InputStatus.END_OF_INPUT` and return `InputStatus.END_OF_INPUT` in that case?

Reply:
Actually it would not happen that case. If both inputs were ended before, then the `StreamTask` would call `context.allActionsCompleted()` in last calling `processInput`, so this calling `processInput` would never happen. 

In other words, if the `processInput` is called, that means at-least one of the inputs is not finished yet.